File: web/lib/cwaClient.js
// 封裝中央氣象署 (CWA) Open Data API 未來1週逐12小時天氣預報 (F-D0047 系列)。
// port 自專案根目錄的 cwa_client.py。用 Next.js 的 fetch cache（10 分鐘）取代
// Python 版一次性 CLI 呼叫；不移植 verify=False 的 TLS workaround，那是本機
// Windows Python/OpenSSL 的問題，Vercel 的 Node 執行環境預期不會重現。

import { WEEKLY_DATASET_ID } from "./locations";

const API_URL_TMPL = (datasetId) =>
  `https://opendata.cwa.gov.tw/api/v1/rest/datastore/${datasetId}`;

const REVALIDATE_SECONDS = 600; // 10 分鐘

export class CwaApiError extends Error {}

async function callApi(datasetId, params) {
  const apiKey = process.env.CWA_API_KEY;
  if (!apiKey) {
    throw new CwaApiError("找不到授權碼，請設定環境變數 CWA_API_KEY。");
  }

  const url = new URL(API_URL_TMPL(datasetId));
  url.searchParams.set("Authorization", apiKey);
  url.searchParams.set("format", "JSON");
  for (const [key, value] of Object.entries(params)) {
    if (value) url.searchParams.set(key, value);
  }

  let resp;
  try {
    resp = await fetch(url.toString(), {
      next: { revalidate: REVALIDATE_SECONDS },
    });
  } catch (exc) {
    throw new CwaApiError(`連線中央氣象署 API 失敗：${exc.message}`);
  }

  if (resp.status === 401) {
    throw new CwaApiError("授權碼 (Authorization key) 無效，請確認 CWA_API_KEY 是否正確。");
  }
  if (!resp.ok) {
    const text = await resp.text();
    throw new CwaApiError(`CWA API 回傳非預期狀態碼 ${resp.status}：${text.slice(0, 200)}`);
  }

  const data = await resp.json();
  if (String(data.success).toLowerCase() !== "true") {
    throw new CwaApiError(`CWA API 回應 success=false：${JSON.stringify(data)}`);
  }
  return data;
}

// 回傳 { town, township }：指定鄉鎮的 Location 物件，以及實際使用的鄉鎮名稱。
export async function fetchWeeklyForecast(city, township) {
  const datasetId = WEEKLY_DATASET_ID[city];
  if (!datasetId) {
    throw new CwaApiError(`「${city}」目前沒有對應的一週預報 dataset id。`);
  }

  const data = await callApi(datasetId, { LocationName: township });
  const townList = data.records?.Locations?.[0]?.Location ?? [];
  if (townList.length === 0) {
    throw new CwaApiError(
      `「${city}」查無鄉鎮「${township}」的資料，請確認鄉鎮市區名稱是否正確。`
    );
  }
  return townList[0];
}

// 回傳該縣市轄下所有鄉鎮市區的 Location 物件陣列（一次 API 呼叫拿全部）。
export async function fetchAllTownships(city) {
  const datasetId = WEEKLY_DATASET_ID[city];
  if (!datasetId) {
    throw new CwaApiError(`「${city}」目前沒有對應的一週預報 dataset id。`);
  }

  const data = await callApi(datasetId, {});
  const townList = data.records?.Locations?.[0]?.Location ?? [];
  if (townList.length === 0) {
    throw new CwaApiError(`「${city}」查無任何鄉鎮資料。`);
  }
  return townList;
}

// 把 WeatherElement 攤平、依 StartTime 對齊，回傳依時間排序的 12 小時預報列表。
export function parseForecast(town) {
  const records = new Map();

  for (const element of town.WeatherElement ?? []) {
    for (const slot of element.Time ?? []) {
      const start = slot.StartTime;
      if (!records.has(start)) {
        records.set(start, { start: new Date(start), end: new Date(slot.EndTime) });
      }
      const record = records.get(start);
      for (const value of slot.ElementValue ?? []) {
        if (value.MaxTemperature !== undefined) record.maxT = toNumber(value.MaxTemperature);
        if (value.MinTemperature !== undefined) record.minT = toNumber(value.MinTemperature);
        if (value.ProbabilityOfPrecipitation !== undefined)
          record.pop = toNumber(value.ProbabilityOfPrecipitation);
        if (value.Weather !== undefined) record.wx = value.Weather;
        if (value.WeatherCode !== undefined) record.wxCode = value.WeatherCode;
        if (value.MaxComfortIndexDescription !== undefined)
          record.ci = value.MaxComfortIndexDescription;
      }
    }
  }

  if (records.size === 0) {
    throw new CwaApiError("解析結果為空，CWA 回應格式可能與預期不同。");
  }

  return [...records.values()].sort((a, b) => a.start - b.start);
}

function toNumber(value) {
  const n = Number(value);
  return Number.isNaN(n) ? null : n;
}

File: cwa_client.py
"""封裝中央氣象署 (CWA) Open Data API 未來1週逐12小時天氣預報 (F-D0047 系列)。"""
from __future__ import annotations

import sys
from datetime import datetime
from typing import Any

import requests
import urllib3
from requests.exceptions import SSLError

import locations

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

API_URL_TMPL = "https://opendata.cwa.gov.tw/api/v1/rest/datastore/{dataset_id}"
REQUEST_TIMEOUT = 20

_warned_about_tls = False


class CwaApiError(RuntimeError):
    """CWA API 呼叫失敗或回傳非預期內容時丟出。"""


def _get(url: str, params: dict[str, str]) -> requests.Response:
    """先用正常憑證驗證連線；若命中已知的 CWA 憑證缺欄位問題才退回不驗證，並警告使用者。"""
    global _warned_about_tls
    try:
        return requests.get(url, params=params, timeout=REQUEST_TIMEOUT)
    except SSLError as exc:
        if "Subject Key Identifier" not in str(exc):
            raise
        if not _warned_about_tls:
            print(
                "警告：opendata.cwa.gov.tw 憑證未通過本機嚴格 TLS 驗證"
                "（Missing Subject Key Identifier）。已用 curl 交叉比對確認該網域可正常"
                "以系統信任的憑證連線，判斷是 Python/OpenSSL 對此憑證的檢查較嚴格所致，"
                "本次呼叫改為略過憑證驗證。若你的網路環境有代理伺服器攔截 HTTPS，"
                "請自行確認風險。",
                file=sys.stderr,
            )
            _warned_about_tls = True
        return requests.get(url, params=params, timeout=REQUEST_TIMEOUT, verify=False)


def fetch_weekly_forecast(city: str, api_key: str, township: str | None = None) -> tuple[dict[str, Any], str]:
    """呼叫該縣市對應的 F-D0047 資料集，回傳 (該鄉鎮的 Location 物件, 實際使用的鄉鎮名稱)。"""
    dataset_id = locations.WEEKLY_DATASET_ID.get(city)
    if dataset_id is None:
        raise CwaApiError(f"「{city}」目前沒有對應的一週預報 dataset id。")

    target_township = township or locations.DEFAULT_TOWNSHIP[city]

    try:
        resp = _get(
            API_URL_TMPL.format(dataset_id=dataset_id),
            {"Authorization": api_key, "format": "JSON", "LocationName": target_township},
        )
    except requests.RequestException as exc:
        raise CwaApiError(f"連線中央氣象署 API 失敗：{exc}") from exc

    if resp.status_code == 401:
        raise CwaApiError("授權碼 (Authorization key) 無效，請確認 --key 或 CWA_API_KEY 是否正確。")
    if resp.status_code != 200:
        raise CwaApiError(f"CWA API 回傳非預期狀態碼 {resp.status_code}：{resp.text[:200]}")

    data = resp.json()
    if str(data.get("success")).lower() != "true":
        raise CwaApiError(f"CWA API 回應 success=false：{data}")

    location_groups = data.get("records", {}).get("Locations", [])
    town_list = location_groups[0].get("Location", []) if location_groups else []
    if not town_list:
        raise CwaApiError(
            f"「{city}」查無鄉鎮「{target_township}」的資料，"
            f"請用 --township 指定{city}轄下正確的鄉鎮市區名稱。"
        )
    return town_list[0], target_township


def fetch_all_townships(city: str, api_key: str) -> list[dict[str, Any]]:
    """呼叫該縣市對應的 F-D0047 資料集，不指定鄉鎮，回傳該縣市轄下所有鄉鎮市區的 Location 物件列表。"""
    dataset_id = locations.WEEKLY_DATASET_ID.get(city)
    if dataset_id is None:
        raise CwaApiError(f"「{city}」目前沒有對應的一週預報 dataset id。")

    try:
        resp = _get(
            API_URL_TMPL.format(dataset_id=dataset_id),
            {"Authorization": api_key, "format": "JSON"},
        )
    except requests.RequestException as exc:
        raise CwaApiError(f"連線中央氣象署 API 失敗：{exc}") from exc

    if resp.status_code == 401:
        raise CwaApiError("授權碼 (Authorization key) 無效，請確認 --key 或 CWA_API_KEY 是否正確。")
    if resp.status_code != 200:
        raise CwaApiError(f"CWA API 回傳非預期狀態碼 {resp.status_code}：{resp.text[:200]}")

    data = resp.json()
    if str(data.get("success")).lower() != "true":
        raise CwaApiError(f"CWA API 回應 success=false：{data}")

    location_groups = data.get("records", {}).get("Locations", [])
    town_list = location_groups[0].get("Location", []) if location_groups else []
    if not town_list:
        raise CwaApiError(f"「{city}」查無任何鄉鎮資料。")
    return town_list


def parse_forecast(town: dict[str, Any]) -> list[dict[str, Any]]:
    """把 WeatherElement 攤平、依 StartTime 對齊，回傳依時間排序的 12 小時預報列表。"""
    records: dict[str, dict[str, Any]] = {}

    for element in town.get("WeatherElement", []):
        for slot in element.get("Time", []):
            start = slot["StartTime"]
            record = records.setdefault(
                start,
                {"start": datetime.fromisoformat(start), "end": datetime.fromisoformat(slot["EndTime"])},
            )
            for value in slot.get("ElementValue", []):
                if "MaxTemperature" in value:
                    record["max_t"] = _to_number(value["MaxTemperature"])
                if "MinTemperature" in value:
                    record["min_t"] = _to_number(value["MinTemperature"])
                if "ProbabilityOfPrecipitation" in value:
                    record["pop"] = _to_number(value["ProbabilityOfPrecipitation"])
                if "Weather" in value:
                    record["wx"] = value["Weather"]
                if "WeatherCode" in value:
                    record["wx_code"] = value["WeatherCode"]
                if "MaxComfortIndexDescription" in value:
                    record["ci"] = value["MaxComfortIndexDescription"]

    if not records:
        raise CwaApiError("解析結果為空，CWA 回應格式可能與預期不同，請檢查原始 JSON。")

    return sorted(records.values(), key=lambda r: r["start"])


def _to_number(value: Any) -> float | None:
    try:
        return float(value)
    except (TypeError, ValueError):
        return None
